Added inferior-julia comint mode

Elzair opened the pull request:
I added a comint-mode for julia. Now, one can run the Julia interpreter in Emacs with `M-x inferior-julia`

Elzair committed:
Added inferior-julia comint mode

Elzair committed:
Changed inferior-julia function in contrib/julia-mode.el to apply
make-comint-in-buffer instead of calling function directly

Elzair committed:
Changed input method for Julia Emacs buffer to TeX

Elzair committed:
Removed (setq comint-proccess-echoes t) from contrib/julia-mode.el

Elzair commented:
I have also changed the input mode for the Julia buffer to TeX, so it can now do TeX formatting.

pao commented:
cc @Wilfred

Wilfred commented on contrib/julia-mode.el:
```diff
@@ -3095,6 +3095,42 @@ end"))
+;; Code for `inferior-julia-mode'
+(require 'comint)
+
+(defvar julia-file-path "/usr/bin/julia" "Path to the program used by `inferior-julia'.")
```
It might be nicer to use `defcustom` here, what do you think? I'd also suggest just using `"julia"` rather than hard-coding the path.

Wilfred commented on contrib/julia-mode.el:
```diff
@@ -3095,6 +3095,42 @@ end"))
+(defvar inferior-julia-prompt-regexp "julia>"
+  "Prompt for `inferior-julia'.")
+
+(defun inferior-julia ()
```
Worth adding an `;;;###autoload` cookie here.

Wilfred commented on contrib/julia-mode.el:
```diff
@@ -3095,6 +3095,42 @@ end"))
+
+(define-derived-mode inferior-julia-mode comint-mode "Julia"
+  "Major mode for `inferior-julia'."
+  nil "Julia"
```
Emacs convention is to generally name interactive buffers with asterisks, so I'd use `*Julia*` here.

Wilfred commented on contrib/julia-mode.el:
```diff
@@ -3095,6 +3095,42 @@ end"))
+  (set (make-local-variable 'font-lock-defaults) '(julia-font-lock-keywords t))
+  (set (make-local-variable 'paragraph-start) inferior-julia-prompt-regexp)
+  (set (make-local-variable 'indent-line-function) 'julia-indent-line)
+  (set-input-method "TeX"))
```
We're trying to avoid TeX input with Julia code, preferring to emulate the interpreter instead. See #8947.

Wilfred commented:
Thanks, this is really good! Left a few minor comments, let me know what you think/whether they make sense :).

Elzair committed:
Changed input from TeX mode to use Julia unicode input

Elzair committed:
Added customization options

Elzair committed:
Changed name of Emacs buffer to *Julia*

Elzair commented:
I like the suggestions. I enabled `julia-mode` unicode input in `inferior-julia-mode`, changed both `julia-file-path` and `julia-arguments` to customizable variables (under the group "Julia" in "Programming->Languages"), removed the absolute file path in `julia-file-path` (renamed to `julia-program`), and finally changed the buffer name to `*Julia*`

Wilfred commented:
It's still probably worth adding an autoload cookie so users can call `inferior-julia` without being required to write `(require 'julia)`.

Anyway, this is really good thing to have in julia-mode.el :+1:

Elzair committed:
Added autoload cookie to inferior-julia function in contrib/julia-mode.el

Elzair commented:
Okay, the autoload cookie has been added

Elzair commented:
Does everything look good?

pao commented:
@Wilfred I'm merging since it looks like this addresses your comments. Feel free to follow up if I've ruined everything.

Wilfred commented:
Marvellous, I was about to ask you to merge it. Thanks :)
